File: 0148-sort-list/0148-sort-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 
 struct ListNode* createNode() {
    struct ListNode *node; //why pointer? malloc 할당을 받기 위해 
    node = (struct ListNode *)malloc(sizeof(struct ListNode));
    return node;
}

//조작이 원본에 반영이 되려면 pointer로 넘겨야 한다 
struct ListNode* mergeList(struct ListNode* left, struct ListNode* right){
    struct ListNode *dummy = createNode();
    struct ListNode *tail = dummy;
        while (left && right)
        {
            if (left->val > right->val)
            {
                tail->next = right;
                right = right->next;
            }
            else
            {
                tail->next = left;
                left = left->next;
            }
            tail = tail->next;
        }
        while (left)
        {
            tail->next = left;
            left = left->next;
            tail = tail->next;
        }
        while (right)
        {
            tail->next = right;
            right = right->next;
            tail = tail->next;
        }
        return dummy->next;
}
struct ListNode* findMid(struct ListNode* head) {
    struct ListNode* low = head;
    struct ListNode* high = head->next; //둘다 pointer로 선언해야 접근 가능하다고 한다. 
    while (high && high->next)
    {
        low = low->next;
        high = high->next->next;
    }
    return low;
}

struct ListNode* sortList(struct ListNode* head) {
    if (head == NULL || head->next == NULL)
        return head;
    struct ListNode* mid = findMid(head);
    struct ListNode* left = head;
    struct ListNode* right = mid->next;
    mid->next = NULL;

    left = sortList(left);
    right = sortList(right);
    return mergeList(left, right);
}

File: 0148-sort-list/0148-sort-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def sortList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if not head or not head.next:
            return head
        
        mid = self.findMid(head)
        left = head
        right = mid.next
        mid.next = None

        left = self.sortList(left)
        right = self.sortList(right)
        return self.mergeList(left, right)

    def findMid(self, head):
        low, high = head, head.next
        while high and high.next:
            low = low.next
            high = high.next.next
        return low
    def mergeList(self, left, right):
        dummy = tail = ListNode()

        while left and right:
            if left.val > right.val:
                #print("left : ", left.val, "right : ", right.val)
                tail.next = right
                right = right.next
            else:
                #print("!!left : ", left.val, "right : ", right.val)
                tail.next = left
                left = left.next
            tail = tail.next
        while left:
            tail.next = left
            left = left.next
            tail = tail.next
        while right:
            tail.next = right
            right = right.next
            tail = tail.next
        return dummy.next
